File: backend/app/models/schemas/organization.py
"""
Copyright 2024-2026 ChatterMate

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

from pydantic import BaseModel, EmailStr, field_validator
from typing import Optional, Dict, TypedDict
from uuid import UUID

from app.core.security import validate_password_strength
from app.models.schemas.user import UserResponse


class BusinessHours(TypedDict):
    start: str
    end: str
    enabled: bool


class BusinessHoursDict(TypedDict):
    monday: BusinessHours
    tuesday: BusinessHours
    wednesday: BusinessHours
    thursday: BusinessHours
    friday: BusinessHours
    saturday: BusinessHours
    sunday: BusinessHours


class OrganizationBase(BaseModel):
    name: str
    domain: str
    timezone: Optional[str] = 'UTC'
    business_hours: Optional[BusinessHoursDict] = {
        'monday': {'start': '09:00', 'end': '17:00', 'enabled': True},
        'tuesday': {'start': '09:00', 'end': '17:00', 'enabled': True},
        'wednesday': {'start': '09:00', 'end': '17:00', 'enabled': True},
        'thursday': {'start': '09:00', 'end': '17:00', 'enabled': True},
        'friday': {'start': '09:00', 'end': '17:00', 'enabled': True},
        'saturday': {'start': '09:00', 'end': '17:00', 'enabled': False},
        'sunday': {'start': '09:00', 'end': '17:00', 'enabled': False}
    }
    settings: Optional[Dict] = {}


class OrganizationCreate(OrganizationBase):
    admin_email: EmailStr
    admin_name: str
    admin_password: str

    @field_validator('admin_password')
    @classmethod
    def _check_strength(cls, value: str) -> str:
        # The owner account created at signup clears the same bar as an invited
        # teammate or a password reset. This field previously had no validation
        # at all, which made the front door the weakest way into the product.
        return validate_password_strength(value)


class OrganizationUpdate(BaseModel):
    name: Optional[str] = None
    domain: Optional[str] = None
    timezone: Optional[str] = None
    business_hours: Optional[BusinessHoursDict] = None
    settings: Optional[Dict] = None


class OrganizationCreateResponse(OrganizationBase):
    id: UUID
    is_active: bool
    access_token: Optional[str] = None
    refresh_token: Optional[str] = None
    token_type: Optional[str] = None
    user: UserResponse
    # True when this deployment blocks unverified sign-ins, in which case no
    # session was issued above and the caller must go and click the link.
    email_verification_required: bool = False
    # Whether the verification mail actually went out. False means SMTP is
    # unconfigured or refused it — the UI needs to say so rather than tell the
    # user to check an inbox nothing was sent to.
    email_verification_sent: bool = False

    class Config:
        from_attributes = True


class OrganizationResponse(OrganizationBase):
    id: UUID
    is_active: bool
    access_token: Optional[str] = None
    refresh_token: Optional[str] = None
    token_type: Optional[str] = None

File: frontend/src/types/organization.ts
/*
Copyright 2024-2026 ChatterMate

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import type { User } from "@/types/user"

export interface BusinessHours {
  start: string
  end: string
  enabled: boolean
}

export interface BusinessHoursDict {
  monday: BusinessHours
  tuesday: BusinessHours
  wednesday: BusinessHours
  thursday: BusinessHours
  friday: BusinessHours
  saturday: BusinessHours
  sunday: BusinessHours
}

export interface Organization {
  id: string
  name: string
  domain: string
  timezone: string
  business_hours: BusinessHoursDict
  settings: Record<string, any>
  is_active: boolean
}

export interface OrganizationCreate {
  name: string
  domain: string
  timezone: string
  business_hours: BusinessHoursDict
  admin_email: string
  admin_name: string
  admin_password: string
  settings?: Record<string, any>
}

export interface OrganizationUpdate {
  name?: string
  domain?: string
  timezone?: string
  business_hours?: BusinessHoursDict
  settings?: Record<string, any>
}

export interface OrganizationResponse extends Omit<OrganizationCreate, 'admin_password'> {
  id: string
  is_active: boolean
  access_token?: string
  refresh_token?: string
  token_type?: string
  user: User
  /** Deployment blocks unverified sign-in, so no session was issued. */
  email_verification_required?: boolean
  /** Whether the verification email actually left the server. */
  email_verification_sent?: boolean
}
